estate/models/estate_property.py

```diff
@@ -0,0 +1,106 @@
+    date_available = fields.Datetime(
+        "Available From",
+        copy=False,
+        default=date.today() + relativedelta(months=3),
+    )
+    postcode = fields.Char()
+    expected_price = fields.Float()
+    selling_price = fields.Float(readonly=True, copy=False)
+    best_price = fields.Float(compute='_compute_best_price', readonly=True)
+    bedrooms = fields.Integer(default=2)
+    facades = fields.Integer()
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    living_area = fields.Integer("Living Area (sqm)")
+    garden_area = fields.Integer()
+    total_area = fields.Float(compute="_compute_total_area", readonly=True)
+    garden_orientation = fields.Selection(
+        selection=[('north', 'North'), ('east', 'East'), ('south', 'South'), ('west', 'West')],
+    )
+    state = fields.Selection(
+        selection=[('new', 'New'), ('offer_received', 'Offer Received'), ('offer_accepted', 'Offer Accepted'),
+                   ('sold', 'Sold'), ('cancelled', 'Cancelled')],
+        required=True,
+        default='new',
+        copy=False
+    )
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_new_or_cancelled(self):
+        if self.state not in ('new', 'cancelled'):
+            raise UserError("Only new or cancelled properties can be deleted")
+
+    @api.constrains('selling_price', 'expected_price')
+    def _check_selling_price(self):
+        for record in self:
+            if record.selling_price == 0:
+                continue
+            if float_compare(record.selling_price, record.expected_price * 0.9, 1) < 0:
+                raise ValidationError("The selling price cannot be lower than 90% of the expected price")
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.garden_area + record.living_area
+
+    @api.depends('offer_ids.price')
+    def _compute_best_price(self):
+        for record in self:
+            record.best_price = max(record.offer_ids.mapped('price'), default=0)
+
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = 'south'
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
+
+    def action_sell(self):
+        for record in self:
+            if record.state == 'cancelled':
+                raise UserError("A cancelled property cannot be sold")
+            if len(record.offer_ids.filtered(lambda offer: offer.status == 'accepted')) == 0:
+                raise UserError("A property without any accepted offers cannot be sold")
+            record.state = 'sold'
+        return True
+
+    def action_cancel(self):
+        for record in self:
+            if record.state == 'sold':
+                raise UserError("A sold property cannot be cancelled")
+            record.state = 'cancelled'
+        return True
```

Comment:
```suggestion
    def action_cancel(self):
        if any(record.state == 'sold' for record in self):
            raise UserError("A sold property cannot be cancelled")
        self.state = 'cancelled'
        return True
```
The power of recordsets 2: electric boogaloo